File: scripts/audit-comprehensive.js
#!/usr/bin/env node

/**
 * Sa Đéc Marketing Hub - Comprehensive Audit Script
 * Quét broken links, meta tags, và accessibility issues
 *
 * Usage: node scripts/audit-comprehensive.js
 */

const fs = require('fs')
const path = require('path')

const ROOT_DIR = path.join(__dirname, '..')
const EXCLUDE_DIRS = ['node_modules', 'dist', 'reports', 'playwright-report', '.git', '.vercel']

// Results storage
const results = {
  summary: {
    totalFiles: 0,
    totalIssues: 0
  },
  brokenLinks: [],
  missingMetaDescription: [],
  missingTitle: [],
  missingOgTags: [],
  missingAltImages: [],
  missingLangAttribute: [],
  emptyLinks: [],
  multipleH1: [],
  skippedHeadingLevels: [],
  missingFavicon: [],
  missingCanonical: [],
  accessibilityIssues: []
}

// Get all HTML files
function getHtmlFiles(dir, fileList = []) {
  const files = fs.readdirSync(dir)

  for (const file of files) {
    const filePath = path.join(dir, file)

    // Skip excluded directories
    if (EXCLUDE_DIRS.some(excluded => filePath.includes(excluded))) {
      continue
    }

    if (fs.statSync(filePath).isDirectory()) {
      getHtmlFiles(filePath, fileList)
    } else if (file.endsWith('.html')) {
      fileList.push(filePath)
    }
  }

  return fileList
}

// Audit a single HTML file
function auditFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8')
  const relativePath = path.relative(ROOT_DIR, filePath)
  const issues = []

  // 1. Check for broken hash links
  const hashLinkRegex = /href="#([^"]+)"/g
  let match
  const validHashIds = [...content.matchAll(/id="([^"]+)"/g)].map(m => m[1])
  const validHashNames = [...content.matchAll(/name="([^"]+)"/g)].map(m => m[1])
  const validHashes = new Set([...validHashIds, ...validHashNames, 'top', 'content', 'main', 'skip'])

  while ((match = hashLinkRegex.exec(content)) !== null) {
    const hash = match[1]
    if (hash && !validHashes.has(hash) && !hash.startsWith('!')) {
      results.brokenLinks.push({ file: relativePath, hash: '#' + hash })
      issues.push(`Broken hash link: #${hash}`)
    }
  }

  // 2. Check for missing meta description
  const hasMetaDesc = /<meta\s+name=["']description["']/i.test(content)
  const hasOgDesc = /<meta\s+property=["']og:description["']/i.test(content)
  if (!hasMetaDesc && !hasOgDesc) {
    results.missingMetaDescription.push(relativePath)
    issues.push('Missing meta description')
  }

  // 3. Check for missing or empty title
  const titleMatch = content.match(/<title>([\s\S]*?)<\/title>/i)
  if (!titleMatch || !titleMatch[1].trim()) {
    results.missingTitle.push(relativePath)
    issues.push('Missing or empty title tag')
  }

  // 4. Check for missing Open Graph tags
  const hasOgTitle = /<meta\s+property=["']og:title["']/i.test(content)
  const hasOgImage = /<meta\s+property=["']og:image["']/i.test(content)
  if (!hasOgTitle || !hasOgImage) {
    results.missingOgTags.push({
      file: relativePath,
      missing: { ogTitle: !hasOgTitle, ogImage: !hasOgImage }
    })
    issues.push(`Missing OG tags: ${!hasOgTitle ? 'og:title ' : ''}${!hasOgImage ? 'og:image' : ''}`)
  }

  // 5. Check for images without alt text
  const imgRegex = /<img\s+([^>]+)>/gi
  let imgMatch
  while ((imgMatch = imgRegex.exec(content)) !== null) {
    const attrs = imgMatch[1]
    const hasAlt = /alt=["']([^"']*)["']/i.test(attrs)
    if (!hasAlt) {
      results.missingAltImages.push({ file: relativePath, snippet: imgMatch[0].slice(0, 80) })
      issues.push('Image missing alt attribute')
    }
  }

  // 6. Check for missing lang attribute on html
  const htmlTag = content.match(/<html\s+([^>]*)>/i)
  if (htmlTag && !/lang=["'][^"']+["']/i.test(htmlTag[1])) {
    results.missingLangAttribute.push(relativePath)
    issues.push('Missing lang attribute on <html>')
  }

  // 7. Check for empty links (no text, no aria-label)
  const linkRegex = /<a\s+([^>]*)>([\s\S]*?)<\/a>/gi
  let linkMatch
  while ((linkMatch = linkRegex.exec(content)) !== null) {
    const attrs = linkMatch[1]
    const text = linkMatch[2].replace(/<[^>]*>/g, '').trim()
    const hasAriaLabel = /aria-label=["'][^"']+["']/i.test(attrs)
    const hasHref = /href=["'][^"']+["']/i.test(attrs)

    if (!text && !hasAriaLabel && hasHref) {
      results.emptyLinks.push({ file: relativePath, snippet: linkMatch[0].slice(0, 80) })
      issues.push('Empty link without aria-label')
    }
  }

  // 8. Check for multiple H1 tags
  const h1Tags = content.match(/<h1[^>]*>/gi) || []
  if (h1Tags.length > 1) {
    results.multipleH1.push({ file: relativePath, count: h1Tags.length })
    issues.push(`Multiple H1 tags (${h1Tags.length})`)
  }

  // 9. Check heading hierarchy
  const headings = []
  const hRegex = /<(h[1-6])[^>]*>(.*?)<\/\1>/gi
  let hMatch
  while ((hMatch = hRegex.exec(content)) !== null) {
    const level = parseInt(hMatch[1].substring(1))
    headings.push({ level, text: hMatch[2].slice(0, 50) })
  }

  let lastLevel = 0
  for (const h of headings) {
    if (h.level > lastLevel + 1 && lastLevel !== 0) {
      results.skippedHeadingLevels.push({
        file: relativePath,
        issue: `Skipped from H${lastLevel} to H${h.level}`
      })
      issues.push(`Skipped heading level: H${lastLevel} → H${h.level}`)
    }
    lastLevel = h.level
  }

  // 10. Check for missing favicon
  if (!/rel=["'](shortcut )?icon["']/i.test(content)) {
    results.missingFavicon.push(relativePath)
    issues.push('Missing favicon')
  }

  // 11. Check for missing canonical URL
  if (!/<link\s+rel=["']canonical["']/i.test(content)) {
    results.missingCanonical.push(relativePath)
    issues.push('Missing canonical URL')
  }

  // Store accessibility issues
  if (issues.length > 0) {
    results.accessibilityIssues.push({
      file: relativePath,
      issues
    })
  }
}

// Print section helper
function printSection(title, items, max = 10) {
  if (items.length === 0) {
    } else {
    `)
    items.slice(0, max).forEach(item => {
      if (typeof item === 'string') {
        } else {
        const detail = item.file ? item.file : 'unknown'
        const extra = item.hash ? ` ${item.hash}` : item.count ? ` (${item.count} tags)` : ''
        }
    })
    if (items.length > max) {
      }
  }
  }

// Main execution
const htmlFiles = getHtmlFiles(ROOT_DIR)
results.summary.totalFiles = htmlFiles.length
for (const file of htmlFiles) {
  auditFile(file)
}

// Calculate total issues
results.summary.totalIssues = Object.values(results).reduce((sum, value) => {
  if (Array.isArray(value)) return sum + value.length
  return sum
}, 0)

// Print results
printSection('Broken Hash Links', results.brokenLinks)
printSection('Missing Meta Description', results.missingMetaDescription)
printSection('Missing Title Tags', results.missingTitle)
printSection('Missing Open Graph Tags', results.missingOgTags)
printSection('Images Missing Alt Text', results.missingAltImages, 5)
printSection('Missing Lang Attribute', results.missingLangAttribute)
printSection('Empty Links', results.emptyLinks, 5)
printSection('Multiple H1 Tags', results.multipleH1)
printSection('Skipped Heading Levels', results.skippedHeadingLevels)
printSection('Missing Favicon', results.missingFavicon)
printSection('Missing Canonical URL', results.missingCanonical)

// Save report
const reportsDir = path.join(ROOT_DIR, 'reports', 'audit')
if (!fs.existsSync(reportsDir)) {
  fs.mkdirSync(reportsDir, { recursive: true })
}

const dateStr = new Date().toISOString().split('T')[0]
const reportPath = path.join(reportsDir, `comprehensive-audit-${dateStr}.json`)
fs.writeFileSync(reportPath, JSON.stringify(results, null, 2))

, reportPath)}`)
// Exit with error code if issues found
if (results.summary.totalIssues > 0) {
  process.exit(1)
} else {
  process.exit(0)
}

File: scripts/audit-comprehensive.py
#!/usr/bin/env python3
"""
Comprehensive Audit Script — Sa Đéc Marketing Hub
Quét: Broken Links, Meta Tags, Accessibility Issues
"""

import os
import re
import sys
from pathlib import Path
from dataclasses import dataclass
from typing import List, Dict, Tuple

@dataclass
class Issue:
    file: str
    line: int
    type: str
    severity: str
    description: str
    suggestion: str

class ComprehensiveAudit:
    def __init__(self, root_dir: str):
        self.root_dir = Path(root_dir)
        self.issues: List[Issue] = []
        self.stats = {
            'files_scanned': 0,
            'broken_links': 0,
            'meta_issues': 0,
            'a11y_issues': 0,
        }
    
    def scan_html_files(self) -> List[Path]:
        """Tìm tất cả file HTML"""
        html_files = []
        for pattern in ['**/*.html', '**/*.htm']:
            html_files.extend(self.root_dir.glob(pattern))
        return html_files
    
    def check_broken_links(self, file: Path, content: str):
        """Kiểm tra broken links"""
        lines = content.split('\n')
        
        # Find all href and src attributes
        link_pattern = re.compile(r'(href|src)\s*=\s*["\']([^"\']+)["\']', re.IGNORECASE)
        
        for line_num, line in enumerate(lines, 1):
            matches = link_pattern.findall(line)
            for attr, link in matches:
                # Skip external links, anchors, and special protocols
                if link.startswith(('http', 'https', '#', 'tel:', 'mailto:', 'javascript:', 'data:')):
                    continue
                
                # Check internal links
                if not link.startswith('/') and not link.startswith('./') and not link.startswith('../'):
                    continue
                
                # Extract path without query params
                link_path = link.split('?')[0].split('#')[0]
                
                # Resolve relative path
                if link_path.startswith('/'):
                    target = self.root_dir / link_path.lstrip('/')
                else:
                    target = file.parent / link_path
                
                # Check if file exists
                if not target.exists() and not target.with_suffix('.html').exists():
                    # Check if it's a valid directory
                    if not target.is_dir():
                        self.issues.append(Issue(
                            file=str(file.relative_to(self.root_dir)),
                            line=line_num,
                            type='broken_link',
                            severity='high',
                            description=f'Link không tồn tại: {link}',
                            suggestion=f'Kiểm tra đường dẫn: {link}'
                        ))
                        self.stats['broken_links'] += 1
    
    def check_meta_tags(self, file: Path, content: str):
        """Kiểm tra meta tags"""
        lines = content.split('\n')
        
        # Required meta tags
        required_meta = {
            'title': False,
            'description': False,
            'viewport': False,
            'charset': False,
            'og:title': False,
            'og:description': False,
            'og:image': False,
            'og:url': False,
            'twitter:card': False,
        }
        
        # Check for presence
        for line in lines:
            line_lower = line.lower()
            if '<title>' in line_lower:
                required_meta['title'] = True
            if 'name="description"' in line_lower:
                required_meta['description'] = True
            if 'name="viewport"' in line_lower:
                required_meta['viewport'] = True
            if 'charset=' in line_lower:
                required_meta['charset'] = True
            if 'property="og:title"' in line_lower:
                required_meta['og:title'] = True
            if 'property="og:description"' in line_lower:
                required_meta['og:description'] = True
            if 'property="og:image"' in line_lower:
                required_meta['og:image'] = True
            if 'property="og:url"' in line_lower:
                required_meta['og:url'] = True
            if 'name="twitter:card"' in line_lower:
                required_meta['twitter:card'] = True
        
        # Report missing meta tags
        rel_path = str(file.relative_to(self.root_dir))
        if not required_meta['title']:
            self.issues.append(Issue(
                file=rel_path,
                line=1,
                type='missing_meta',
                severity='medium',
                description='Thiếu thẻ <title>',
                suggestion='Thêm <title>Mô tả trang</title>'
            ))
            self.stats['meta_issues'] += 1
        
        if not required_meta['description']:
            self.issues.append(Issue(
                file=rel_path,
                line=1,
                type='missing_meta',
                severity='medium',
                description='Thiếu meta description',
                suggestion='Thêm <meta name="description" content="...">'
            ))
            self.stats['meta_issues'] += 1
        
        if not required_meta['viewport']:
            self.issues.append(Issue(
                file=rel_path,
                line=1,
                type='missing_meta',
                severity='medium',
                description='Thiếu viewport meta tag',
                suggestion='Thêm <meta name="viewport" content="width=device-width, initial-scale=1.0">'
            ))
            self.stats['meta_issues'] += 1
    
    def check_accessibility(self, file: Path, content: str):
        """Kiểm tra accessibility issues"""
        lines = content.split('\n')
        rel_path = str(file.relative_to(self.root_dir))
        
        for line_num, line in enumerate(lines, 1):
            # Check for images without alt
            if '<img' in line.lower() and 'alt=' not in line.lower():
                self.issues.append(Issue(
                    file=rel_path,
                    line=line_num,
                    type='a11y',
                    severity='high',
                    description='Image thiếu alt attribute',
                    suggestion='Thêm alt="mô tả image"'
                ))
                self.stats['a11y_issues'] += 1
            
            # Check for buttons without aria-label or text
            if '<button' in line.lower():
                has_text = bool(re.search(r'<button[^>]*>.*?</button>', line, re.IGNORECASE | re.DOTALL))
                has_aria = 'aria-label=' in line.lower()
                has_title = 'title=' in line.lower()
                if not has_text and not has_aria and not has_title:
                    self.issues.append(Issue(
                        file=rel_path,
                        line=line_num,
                        type='a11y',
                        severity='medium',
                        description='Button không có text hoặc aria-label',
                        suggestion='Thêm text content hoặc aria-label'
                    ))
                    self.stats['a11y_issues'] += 1
            
            # Check for links without href
            if '<a ' in line.lower() and 'href=' not in line.lower():
                self.issues.append(Issue(
                    file=rel_path,
                    line=line_num,
                    type='a11y',
                    severity='medium',
                    description='Link thiếu href attribute',
                    suggestion='Thêm href="#" hoặc dùng button'
                ))
                self.stats['a11y_issues'] += 1
            
            # Check for form inputs without labels
            if '<input' in line.lower() and 'type="hidden"' not in line.lower():
                has_id = 'id=' in line.lower()
                has_aria = 'aria-label=' in line.lower() or 'aria-labelledby=' in line.lower()
                if not has_id and not has_aria:
                    self.issues.append(Issue(
                        file=rel_path,
                        line=line_num,
                        type='a11y',
                        severity='medium',
                        description='Input không có id hoặc aria-label',
                        suggestion='Thêm id và label, hoặc aria-label'
                    ))
                    self.stats['a11y_issues'] += 1
            
            # Check for lang attribute on html
            if line_num == 1 or line_num <= 3:
                if '<html' in line.lower() and 'lang=' not in line.lower():
                    self.issues.append(Issue(
                        file=rel_path,
                        line=line_num,
                        type='a11y',
                        severity='high',
                        description='HTML thiếu lang attribute',
                        suggestion='Thêm lang="vi" vào thẻ html'
                    ))
                    self.stats['a11y_issues'] += 1
    
    def run(self) -> Dict:
        """Chạy audit toàn diện"""
        html_files = self.scan_html_files()
        self.stats['files_scanned'] = len(html_files)
        
        for file in html_files:
            try:
                content = file.read_text(encoding='utf-8')
                self.check_broken_links(file, content)
                self.check_meta_tags(file, content)
                self.check_accessibility(file, content)
            except Exception as e:
                print(f"Error processing {file}: {e}")
        
        return {
            'stats': self.stats,
            'issues': self.issues,
            'summary': self.generate_summary()
        }
    
    def generate_summary(self) -> str:
        """Tạo summary report"""
        summary = []
        summary.append("=" * 60)
        summary.append("COMPREHENSIVE AUDIT REPORT — Sa Đéc Marketing Hub")
        summary.append("=" * 60)
        summary.append("")
        summary.append(f"Files scanned: {self.stats['files_scanned']}")
        summary.append("")
        summary.append("ISSUES BY TYPE:")
        summary.append(f"  Broken Links:    {self.stats['broken_links']}")
        summary.append(f"  Meta Issues:     {self.stats['meta_issues']}")
        summary.append(f"  A11y Issues:     {self.stats['a11y_issues']}")
        summary.append(f"  TOTAL:           {len(self.issues)}")
        summary.append("")
        
        # Group by severity
        high = len([i for i in self.issues if i.severity == 'high'])
        medium = len([i for i in self.issues if i.severity == 'medium'])
        low = len([i for i in self.issues if i.severity == 'low'])
        
        summary.append("ISSUES BY SEVERITY:")
        summary.append(f"  High:   {high}")
        summary.append(f"  Medium: {medium}")
        summary.append(f"  Low:    {low}")
        summary.append("")
        
        if self.issues:
            summary.append("DETAILED ISSUES:")
            summary.append("-" * 60)
            for issue in self.issues[:50]:  # First 50 issues
                summary.append(f"[{issue.severity.upper()}] {issue.file}:{issue.line}")
                summary.append(f"  Type: {issue.type}")
                summary.append(f"  Issue: {issue.description}")
                summary.append(f"  Fix: {issue.suggestion}")
                summary.append("")
        
        return "\n".join(summary)

if __name__ == '__main__':
    root = '/Users/mac/mekong-cli/apps/sadec-marketing-hub'
    audit = ComprehensiveAudit(root)
    result = audit.run()
    print(result['summary'])
    
    # Save detailed report
    report_file = Path(root) / 'reports' / 'audit' / 'comprehensive-audit-detailed.md'
    report_file.parent.mkdir(parents=True, exist_ok=True)
    
    with open(report_file, 'w', encoding='utf-8') as f:
        f.write("# Comprehensive Audit Report\n\n")
        f.write(f"**Date:** {__import__('datetime').datetime.now().strftime('%Y-%m-%d')}\n\n")
        f.write("## Summary\n\n")
        f.write(f"- Files scanned: {result['stats']['files_scanned']}\n")
        f.write(f"- Broken links: {result['stats']['broken_links']}\n")
        f.write(f"- Meta issues: {result['stats']['meta_issues']}\n")
        f.write(f"- A11y issues: {result['stats']['a11y_issues']}\n\n")
        f.write("## Issues by Severity\n\n")
        high = len([i for i in result['issues'] if i.severity == 'high'])
        medium = len([i for i in result['issues'] if i.severity == 'medium'])
        low = len([i for i in result['issues'] if i.severity == 'low'])
        f.write(f"- High: {high}\n")
        f.write(f"- Medium: {medium}\n")
        f.write(f"- Low: {low}\n\n")
        
        f.write("## Detailed Issues\n\n")
        for issue in result['issues']:
            f.write(f"### [{issue.severity.upper()}] {issue.file}:{issue.line}\n\n")
            f.write(f"- **Type:** {issue.type}\n")
            f.write(f"- **Issue:** {issue.description}\n")
            f.write(f"- **Suggestion:** {issue.suggestion}\n\n")
    
    print(f"\nDetailed report saved to: {report_file}")
